### PR TITLE
feat: jfrog plugin + helm upload handler

### PR DESCRIPTION
**Pull request by jakobmoellerdev**
#### What this PR does / why we need it

Implements a JFrog Plugin that contains an uploader that is able to upload to JFrog Helm Chart Repositories (different from OCI).

Example Reference Upload configuration:

```yaml
- type: credentials.config.ocm.software
  consumers:
  - identity:
      type: JFrogHelm
      hostname: int.repositories.cloud.sap
    credentials:
    - type: Credentials/v1
      properties:
        username: "your-user-here"
        password: "your-token-here"
- type: uploader.ocm.config.ocm.software
  registrations:
  - name: plugin/jfrog/JFrogHelm
    artifactType: helmChart
    priority: 200
    config:
      type: JFrogHelm/v1alpha1
      url: "your-artifactory-url-here"
      repository: "your-repository-here"
      # reindexAfterUpload: true # in case you want to force a reindex, requires admin repository credentials, off by default
```

The plugin will be installable with 

`ocm install plugin ghcr.io/open-component-model/ocm//ocm.software/plugins/jfrogplugin` 

as it is added to our parallel build list.

Since during development (while this PR does not get merged and we dont have an RC) you cannot use this command, you can choose to either push your own version of the plugin (there are commands in the new makefile), or you can run (also in the new makefile):

```
make -C components/jfrogplugin install
```

which will do a build and install locally so you can use it without fetching it remotely

Once added, you can run something like

```
ocm plugin get jfrog -oyaml
---
element:
  description: "ALPHA GRADE plugin providing custom functions related to interacting
    with JFrog Repositories (e.g. Artifactory).\n\nThis plugin is solely for interacting
    with JFrog Servers and cannot be used for generic repository types.\nThus, you
    should only consider this plugin if\n- You need to use a JFrog specific API\n-
    You cannot use any of the generic (non-jfrog) implementations.\n\nExamples:\n\nYou
    can configure the JFrog plugin as an Uploader in an ocm config file with:\n\n-
    type: uploader.ocm..config.ocm.software\n  registrations:\n  - name: plugin/jfrog/JFrogHelm\n
    \   artifactType: helmChart\n    priority: 200 # must be > 100 to be used over
    the default handler\n    config:\n      type: JFrogHelm/v1alpha1\n      # this
    is only a sample JFrog Server URL, do NOT append /artifactory\n      url: int.repositories.ocm.software
    \n      repository: ocm-helm-test\n"
  forwardLogging: true
  pluginName: jfrog
  pluginVersion: 0.20.0-dev+962ef1469035fbd7b855dff1ccb6ddfc06269745
  shortDescription: jfrog plugin
  uploaders:
  - constraints:
    - artifactType: helmChart
      contextType: ""
      mediaType: ""
      repositoryType: ""
    description: upload artifacts to JFrog HELM repositories by using the JFrog REST
      API.
    name: JFrogHelm
  version: v1
```

to introspect it.

The plugin is now able to be used by OCM.

The plugin registers itself for the mediaTypes of a Helm Chart TGZ as well as OCI artifacts to convert them. Notably, the OCI artifact conversion is lossy because the provenance data is omitted, so back and forth conversion while maintaining digests might not always be possible in a fully trusted environment

#### Which issue(s) this PR fixes

fix https://github.com/open-component-model/ocm/issues/1116

### CODE REVIEW

File: cmds/jfrogplugin/uploaders/helm/helm.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+	requestURL := path.Join(spec.URL, "artifactory", spec.Repository, fmt.Sprintf("%s-%s.tgz", spec.ChartName, spec.ChartVersion))
+	requestURLParsed, err := parseURLAllowNoScheme(requestURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse full request URL: %w", err)
+	}
+	return requestURLParsed, nil
+}
+
+// parseURLAllowNoScheme is an adaptation / hack on url.Parse because
+// url.Parse does not support parsing a URL without a prefixed scheme.
```
Comment (Skarlso) [inline cmds/jfrogplugin/uploaders/helm/helm.go:229]: What? It totally does.

```go
	parsedURL, err := url.Parse("github.com")
	if err != nil {
		return nil, err
	}

        // out -> github.com
	fmt.Println(parsedURL)
```

File: cmds/jfrogplugin/uploaders/helm/helm.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	if purl.Port() != "" {
+		info.ConsumerId.SetNonEmptyValue(ID_PORT, purl.Port())
+	}
+
+	return &info, nil
+}
+
+// EnsureSpecWithHelpFromHint introspects the hint and fills the target spec based on it.
+// It makes sure that the spec can be used to access a JFrog Artifactory HELM Repository.
+func EnsureSpecWithHelpFromHint(spec *JFrogHelmUploaderSpec, hint string) error {
```
Comment (Skarlso) [inline cmds/jfrogplugin/uploaders/helm/helm.go:181]: So I was wondering. Why not make the thing that you want part of the spec instead of using the hint as an extra information? Like, it could be part of the `Config` you are constructing, unless I completely misunderstand how the plug works. 🤔

File: cmds/jfrogplugin/uploaders/helm/helm.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	if refFromHint, err := reference.Parse(hint); err == nil {
+		if refFromHint.Digest() != "" && refFromHint.Object == "" {
+			return fmt.Errorf("the hint contained a valid reference but it was a digest, so it cannot be used to deduce a version of the helm chart: %s", refFromHint)
+		}
+		if spec.ChartName == "" {
+			spec.ChartVersion = refFromHint.Object
+		}
```
Comment (Skarlso) [inline cmds/jfrogplugin/uploaders/helm/helm.go:188]:
```suggestion
		if spec.ChartVersion == "" {
			spec.ChartVersion = refFromHint.Object
		}
```

File: cmds/jfrogplugin/uploaders/helm/upload.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	SetHeadersFromCredentials(req, creds)
+
+	if res, err = client.Do(req); err != nil {
+		return nil, fmt.Errorf("failed to store blob in artifactory: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, res.Body.Close())
+	}()
+
+	if invalid := 200 > res.StatusCode || res.StatusCode >= 300; invalid {
```
Comment (Skarlso) [inline cmds/jfrogplugin/uploaders/helm/upload.go:43]: 302 is redirect that can sometimes happen with proxies. Are we concerned about that at all?

**Review by Skarlso — COMMENTED** [review]
Initial review before starting to play around with this.

File: cmds/jfrogplugin/uploaders/helm/helm.go (hunk 2 above)
Comment (jakobmoellerdev) [reply cmds/jfrogplugin/uploaders/helm/helm.go:181]: Unfortunately that doesnt really work for dynamic repos. 

Assume you want to have a config like in the example I provided in the PR description. In this case the repo name and chart name + version are determined dynamically. If you would put them into the spec instead of the hint, every single helm chart would get transferred to the name+version that is specified in the spec. The hint however is dynamic, and can change for every single artifact.

File: cmds/jfrogplugin/uploaders/helm/upload.go (hunk 4 above)
Comment (jakobmoellerdev) [reply cmds/jfrogplugin/uploaders/helm/upload.go:43]: the default http transport should already follow redirects. I believe this should cover this scenario: https://golang.org/pkg/net/http/#Client

File: cmds/jfrogplugin/uploaders/helm/helm.go (hunk 1 above)
Comment (jakobmoellerdev) [reply cmds/jfrogplugin/uploaders/helm/helm.go:229]: Technically youre correct that it does follow that on all strict semantic urls, however host:port combinations that are not a URL do not work. Example:

```go
package main

import (
	"fmt"
	"net/url"
	"os"
)

func main() {
	for _, tc := range []string {
		"github.com",
		"//www.google.com/support/websearch/?source=g&amp;hl=en",
		"localhost",
		"127.0.0.1",
		"127.0.0.1:8080",
		"dummy://127.0.0.1:8080",
		"localhost:8080",
		"dummy://localhost:8080",
		"https://demo.jfrog.ocm.software/artifactory/my-charts/podinfo-0.0.1.tgz",
		"demo.jfrog.ocm.software/artifactory/my-charts/podinfo-0.0.1.tgz",
	} {
		url, err := url.Parse(tc)
		if err != nil {
			fmt.Fprintln(os.Stderr, err.Error())
			continue
		}
		fmt.Printf("URL: %s\n Scheme: %s\n", url, url.Scheme)
	}
}
```

results in 
```
parse "127.0.0.1:8080": first path segment in URL cannot contain colon
URL: github.com
 Scheme: 
URL: //www.google.com/support/websearch/?source=g&amp;hl=en
 Scheme: 
URL: localhost
 Scheme: 
URL: 127.0.0.1
 Scheme: 
URL: dummy://127.0.0.1:8080
 Scheme: dummy
URL: localhost:8080
 Scheme: localhost
URL: dummy://localhost:8080
 Scheme: dummy
URL: https://demo.jfrog.ocm.software/artifactory/my-charts/podinfo-0.0.1.tgz
 Scheme: https
URL: demo.jfrog.ocm.software/artifactory/my-charts/podinfo-0.0.1.tgz
 Scheme: 
```

as you can see, some ip combinations as well as local host are not technically URLs so they will be incorrectly identified. As such the scheme defaulting might fail, which is why this code exists. If you think those cases are not relevant im the first to vote for removal of that code

File: api/ocm/plugin/cache/updater.go [hunk 5]
```diff
@@ -296,44 +296,57 @@ func (o *PluginUpdater) download(session ocm.Session, cv ocm.ComponentVersionAcc
...
+			return err
+		}
+		defer lock.Close()
+
+		target := filepath.Join(dir, desc.PluginName)
+
+		verb := "installing"
+		if ok, _ := vfs.FileExists(fs, target); ok {
+			if !o.Force && (cv.GetVersion() == o.Current || !o.UpdateMode) {
+				return fmt.Errorf("plugin %s already found in %s", desc.PluginName, dir)
```
[the remark this answers is not in the document]
Comment (jakobmoellerdev) [reply api/ocm/plugin/cache/updater.go:326]: there is an update mode that is set so its basically a force override, because you explicitly install one plugin at a time. 

TLDR: Ultimately its just a design design made before and I didnt want to change the error behavior now.

File: api/ocm/plugin/plugin.go [hunk 6]
```diff
@@ -112,11 +114,32 @@ func (p *pluginImpl) Exec(r io.Reader, w io.Writer, args ...string) (result []by
 		args = append([]string{"--" + ppi.OptPlugingLogConfig, string(data)}, args...)
 	}
 
-	if len(p.config) == 0 {
-		p.ctx.Logger(TAG).Debug("execute plugin action", "path", p.Path(), "args", args)
-	} else {
-		p.ctx.Logger(TAG).Debug("execute plugin action", "path", p.Path(), "args", args, "config", p.config)
+	if p.ctx.Logger(TAG).Enabled(mlog.DebugLevel) {
```
[the remark this answers is not in the document]
Comment (Skarlso) [reply api/ocm/plugin/plugin.go:117]: Ugh, honestly, it would be in the logging library. Something like, WithObfuscation or something like that, but I do not wish to do that here.

File: api/ocm/plugin/plugin.go [hunk 7]
```diff
@@ -112,11 +114,32 @@ func (p *pluginImpl) Exec(r io.Reader, w io.Writer, args ...string) (result []by
...
+		// Not totally safe, but better than nothing.
+		logargs := make([]string, len(args))
+		for i, arg := range args {
+			if logargs[i] != "" {
+				continue
+			}
+			if strings.Contains(arg, "credentials") {
+				if strings.Contains(arg, "=") {
+					logargs[i] = "***"
+				} else if i < len(args)-1 {
```
[the remark this answers is not in the document]
Comment (Skarlso) [reply api/ocm/plugin/plugin.go:129]: Gotcha. The `+1` fix is still relevant though. :D

File: cmds/jfrogplugin/uploaders/helm/helm.go [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+	// OPTIONAL: If not set, defaulted from the passed Hint.
+	ChartName string `json:"name"`
+	// Version is the desired version of the chart
+	// OPTIONAL: If not set, defaulted from the passed Hint.
+	ChartVersion string `json:"version"`
+
+	// Timeout is the maximum duration the upload of the chart can take
+	// before aborting and failing.
+	// OPTIONAL: If not set, set to the internal DEFAULT_TIMEOUT.
+	Timeout time.Duration `json:"timeout"`
```
[the remark this answers is not in the document]
Comment (Skarlso) [reply cmds/jfrogplugin/uploaders/helm/helm.go]: Yah, but what I'm saying is what if someone wants to have `0` as timeout?

File: cmds/jfrogplugin/uploaders/helm/helm.go (hunk 2 above)
Comment (Skarlso) [reply cmds/jfrogplugin/uploaders/helm/helm.go:185]: Ah, I get it. Thanks for the explanation.

File: cmds/jfrogplugin/uploaders/helm/helm.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jakobmoellerdev) [reply cmds/jfrogplugin/uploaders/helm/helm.go]: gotcha... why would someone want that? Do you think disabling timeouts is relevant? I will change to pointer with that in mind.

[the remark this answers is not in the document]
Comment (Skarlso) [reply cmds/jfrogplugin/uploaders/helm/helm.go]: Unsure. Normally, people who set 0 as a timeout either have a very shitty network connection and expect the operation to _take a long time_ or people who upload large data and have no idea how long it would take and don't want to bother with figuring out a timeout value.

[the remark this answers is not in the document]
Comment (jakobmoellerdev) [reply cmds/jfrogplugin/uploaders/helm/helm.go]: adjusted to nil value

File: api/ocm/plugin/plugin.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jakobmoellerdev) [reply api/ocm/plugin/plugin.go]: corrected

File: cmds/jfrogplugin/main.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      repository: ocm-helm-test
+`)
+	p.SetConfigParser(GetConfig)
+
+	u := helm.New()
+	if err := p.RegisterUploader(artifacttypes.HELM_CHART, "", u); err != nil {
+		panic(err)
+	}
+	err := cmds.NewPluginCommand(p).Execute(os.Args[1:])
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (jakobmoellerdev) [reply cmds/jfrogplugin/main.go:54]: added a print to stderr

**Review by Skarlso — APPROVED** [review]
Well done! :) Thanks for bearing with me. :D

**Comment by hilmarf** [comment]
There is one bug: https://github.com/open-component-model/ocm/issues/1063 - `~/.ocm/plugins: no such file or directory`
